modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/TableRowSkinTest.java

```diff
@@ -0,0 +1,158 @@
+        TableColumn<Person, Integer> ageCol = new TableColumn<>("Age");
+        ageCol.setCellValueFactory(new PropertyValueFactory<>("age"));
+
+        tableView.getColumns().addAll(firstNameCol, lastNameCol, emailCol, ageCol);
+
+        ObservableList<Person> items = FXCollections.observableArrayList(
+                new Person("firstName1", "lastName1","email1@javafx.com",1),
+                new Person("firstName2", "lastName2","email2@javafx.com",2),
+                new Person("firstName3", "lastName3","email3@javafx.com",3),
+                new Person("firstName4", "lastName4","email4@javafx.com",4)
+        );
+
+        tableView.setItems(items);
+
+        stageLoader = new StageLoader(tableView);
+    }
+
+    @Test
+    public void removedColumnsShouldRemoveCorrespondingCellsInRowFixedCellSize() {
+        removedColumnsShouldRemoveCorrespondingCellsInRow(true);
+    }
+
+    @Test
+    public void removedColumnsShouldRemoveCorrespondingCellsInRow() {
+        removedColumnsShouldRemoveCorrespondingCellsInRow(false);
+    }
+
+    @Test
+    public void invisibleColumnsShouldRemoveCorrespondingCellsInRowFixedCellSize() {
+        invisibleColumnsShouldRemoveCorrespondingCellsInRow(true);
+    }
+
+    @Test
+    public void invisibleColumnsShouldRemoveCorrespondingCellsInRow() {
+        invisibleColumnsShouldRemoveCorrespondingCellsInRow(false);
+    }
+
+    @After
+    public void after() {
+        stageLoader.dispose();
+    }
+
+    private void invisibleColumnsShouldRemoveCorrespondingCellsInRow(boolean useFixedCellSize) {
+        TableRow<Person> tableRow = setRowFactoryVerifyReturnFirstRow(useFixedCellSize);
+
+        // Set the last 2 columns invisible.
+        tableView.getColumns().get(tableView.getColumns().size() - 1).setVisible(false);
+        tableView.getColumns().get(tableView.getColumns().size() - 2).setVisible(false);
+
+        Toolkit.getToolkit().firePulse();
+
+        // We set 2 columns to invisible, so the cell count should be decremented by 2 as well.
+        assertEquals(tableView.getColumns().size() - 2, tableRow.getChildrenUnmodifiable().size());
+    }
+
+    private void removedColumnsShouldRemoveCorrespondingCellsInRow(boolean useFixedCellSize) {
+        TableRow<Person> tableRow = setRowFactoryVerifyReturnFirstRow(useFixedCellSize);
+
+        // Remove the last 2 columns.
+        tableView.getColumns().remove(tableView.getColumns().size() - 1);
+        tableView.getColumns().remove(tableView.getColumns().size() - 1);
+
+        Toolkit.getToolkit().firePulse();
+
+        // We removed 2 columns, so the cell count should be decremented by 2 as well.
+        assertEquals(tableView.getColumns().size(), tableRow.getChildrenUnmodifiable().size());
+    }
+
+    private TableRow<Person> setRowFactoryVerifyReturnFirstRow(boolean useFixedCellSize) {
+        // We save the first table row to check it later.
+        AtomicReference<TableRow<Person>> tableRowRef = new AtomicReference<>();
+
```

Comment:
wondering a bit about this complicated test setup .. are you aware of the VirtualFlowTestUtils (in test.something.infrastructure)? Using it, a test would shring down to something like:

    @Test
    public void testRemoveColumnsFixed() {
        tableView.setFixedCellSize(20);
        tableView.getColumns().remove(0, 2);
        Toolkit.getToolkit().firePulse();
        assertEquals(tableView.getVisibleLeafColumns().size(), 
                VirtualFlowTestUtils.getCell(tableView, 0).getChildrenUnmodifiable().size());
    }

Or what am I missing?

Reply:
Nice catch! I tried it out and it works! And indeed the code looks much better now.
To be fair, I had a brief look at VirtualFlowTestUtils, as other table/cell tests uses it. Next time I should take a closer look. :P 
I pushed your suggested change.